### PR TITLE
Change the complexity of tokens lookup from linear to log(n)

### PR DESCRIPTION
**Pull request by hhugo**
following #1515 

|  | Time |
| --- | --- |
| before 1515 | 0m0.968s |
| after 1515 | 0m2.993s |
| new |  0m0.895s |

### CODE REVIEW

**Review by gpetiot — APPROVED** [review]
Looks good to me, thanks!

**Comment by emillon** [comment]
Thanks for the optimization. This seems to work but I'm not completely comfortable with having a custom binary search here. Do you think this can be made to work with `Array.binary_search` or using a Set to do the search (this compares start and end pos so the latter might be tricky)?

**Comment by hhugo** [comment]
Now using `Array.binary_search`

**Review by gpetiot — APPROVED** [review]

File: lib/Source.ml [hunk 1]
```diff
@@ -12,13 +12,13 @@
...
 (** Concrete syntax. *)
-type t = {text: string; tokens: (Parser.token * Location.t) list}
+type t = {text: string; tokens: (Parser.token * Location.t) array}
 
 let create text =
   let lexbuf = Lexing.from_string text in
   let rec loop acc =
     match Lexer.token lexbuf with
-    | Parser.EOF -> List.rev acc
+    | Parser.EOF -> Array.of_list (List.rev acc)
```
Comment (emillon) [inline lib/Source.ml:21]:
```suggestion
    | Parser.EOF -> Array.of_list_rev acc
```

**Review by emillon — COMMENTED** [review]
Nice, thanks!

**Pull request merged**